ssl/quic/quic_lcidm.c

```diff
@@ -0,0 +1,454 @@
+        return NULL;
+
+    if ((conn->lcids = lh_QUIC_LCID_new(lcid_hash, lcid_comp)) == NULL) {
+        OPENSSL_free(conn);
+        return NULL;
+    }
+
+    conn->opaque = opaque;
+    lh_QUIC_LCIDM_CONN_insert(lcidm->conns, conn);
+    return conn;
+}
+
+static void lcidm_delete_conn_lcid(QUIC_LCIDM *lcidm, QUIC_LCID *lcid)
+{
+    lh_QUIC_LCID_delete(lcidm->lcids, lcid);
+    lh_QUIC_LCID_delete(lcid->conn->lcids, lcid);
+    --lcid->conn->num_active_lcid;
+    OPENSSL_free(lcid);
+}
+
+/* doall_arg wrapper */
+static void lcidm_delete_conn_lcid_(QUIC_LCID *lcid, void *arg)
+{
+    lcidm_delete_conn_lcid((QUIC_LCIDM *)arg, lcid);
+}
+
+static void lcidm_delete_conn(QUIC_LCIDM *lcidm, QUIC_LCIDM_CONN *conn)
+{
+    lh_QUIC_LCID_doall_arg(conn->lcids, lcidm_delete_conn_lcid_, lcidm);
+    lh_QUIC_LCIDM_CONN_delete(lcidm->conns, conn);
+    lh_QUIC_LCID_free(conn->lcids);
+    OPENSSL_free(conn);
+}
+
+static QUIC_LCID *lcidm_conn_new_lcid(QUIC_LCIDM *lcidm, QUIC_LCIDM_CONN *conn,
+                                      const QUIC_CONN_ID *lcid)
+{
+    QUIC_LCID *lcid_obj;
+
+    if ((lcid_obj = OPENSSL_zalloc(sizeof(*lcid_obj))) == NULL)
+        return NULL;
+
+    lcid_obj->cid = *lcid;
+    lcid_obj->conn = conn;
+    lh_QUIC_LCID_insert(conn->lcids, lcid_obj);
+    lh_QUIC_LCID_insert(lcidm->lcids, lcid_obj);
+    ++conn->num_active_lcid;
+    return lcid_obj;
+}
+
+size_t ossl_quic_lcidm_get_lcid_len(const QUIC_LCIDM *lcidm)
+{
+    return lcidm->lcid_len;
+}
+
+size_t ossl_quic_lcidm_get_num_active_lcid(const QUIC_LCIDM *lcidm,
+                                           void *opaque)
+{
+    QUIC_LCIDM_CONN *conn;
+
+    conn = lcidm_get_conn(lcidm, opaque);
+    if (conn == NULL)
+        return 0;
+
+    return conn->num_active_lcid;
+}
+
+static int gen_rand_conn_id(OSSL_LIB_CTX *libctx, size_t len, QUIC_CONN_ID *cid)
+{
+    if (len > QUIC_MAX_CONN_ID_LEN)
+        return 0;
+
+    cid->id_len = (unsigned char)len;
+
+    if (RAND_bytes_ex(libctx, cid->id, len, len * 8) != 1) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_RAND_LIB);
+        cid->id_len = 0;
+        return 0;
+    }
+
+    return 1;
+}
+
+static int lcidm_generate_cid(QUIC_LCIDM *lcidm,
+                              QUIC_CONN_ID *cid)
+{
+    return gen_rand_conn_id(lcidm->libctx, lcidm->lcid_len, cid);
+}
+
+static int lcidm_generate(QUIC_LCIDM *lcidm,
+                          void *opaque,
+                          unsigned int type,
+                          QUIC_CONN_ID *lcid_out,
+                          uint64_t *seq_num)
+{
+    QUIC_LCIDM_CONN *conn;
+    QUIC_LCID key, *lcid_obj;
+
+    if ((conn = lcidm_upsert_conn(lcidm, opaque)) == NULL)
+        return 0;
+
+    if ((type == LCID_TYPE_INITIAL && conn->next_seq_num > 0)
+        || conn->next_seq_num > OSSL_QUIC_VLINT_MAX)
+        return 0;
+
+    if (!lcidm_generate_cid(lcidm, lcid_out))
+        return 0;
+
+    key.cid = *lcid_out;
+    if (lh_QUIC_LCID_retrieve(lcidm->lcids, &key) != NULL)
```

Comment:
Is this prone to transient failure? i.e. if the dbrng just happens to produce a 20 byte bit pattern that matches a connection id already in the hash table.  Its not very likely of course, but in the event it does happen, should this actually be a retry, rather than a failure (at least not an immediate failure)?